fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/sorting/RecordQueryDamPlan.java

```diff
@@ -0,0 +1,216 @@
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.record.query.plan.AvailableFields;
+import com.apple.foundationdb.record.query.plan.cascades.AliasMap;
+import com.apple.foundationdb.record.query.plan.cascades.CorrelationIdentifier;
+import com.apple.foundationdb.record.query.plan.cascades.GroupExpressionRef;
+import com.apple.foundationdb.record.query.plan.cascades.Quantifier;
+import com.apple.foundationdb.record.query.plan.cascades.TranslationMap;
+import com.apple.foundationdb.record.query.plan.cascades.explain.NodeInfo;
+import com.apple.foundationdb.record.query.plan.cascades.explain.PlannerGraph;
+import com.apple.foundationdb.record.query.plan.cascades.expressions.RelationalExpression;
+import com.apple.foundationdb.record.query.plan.cascades.values.QueriedValue;
+import com.apple.foundationdb.record.query.plan.cascades.values.Value;
+import com.apple.foundationdb.record.query.plan.plans.QueryResult;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlanWithChild;
+import com.apple.foundationdb.record.sorting.MemorySortCursor;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * A query plan implementing sorting in-memory, possibly spilling to disk.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class RecordQueryDamPlan implements RecordQueryPlanWithChild {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Record-Query-Dam-Plan");
+
+    @Nonnull
+    private final Quantifier.Physical inner;
+    @Nonnull
+    private final RecordQuerySortKey key;
+
+    public RecordQueryDamPlan(@Nonnull RecordQueryPlan plan, @Nonnull RecordQuerySortKey key) {
+        this(Quantifier.physical(GroupExpressionRef.of(plan)), key);
+    }
+
+    private RecordQueryDamPlan(@Nonnull Quantifier.Physical inner, @Nonnull RecordQuerySortKey key) {
+        this.inner = inner;
+        this.key = key;
+    }
+
+    @Nonnull
+    @Override
+    @SuppressWarnings({"PMD.CloseResource", "resource"})
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull FDBRecordStoreBase<M> store,
+                                                                     @Nonnull EvaluationContext context,
+                                                                     @Nullable byte[] continuation,
+                                                                     @Nonnull ExecuteProperties executeProperties) {
+        // Since we are sorting, we need to feed through everything from the inner plan,
+        // even just to get the top few.
```

Comment:
Is this comment about sorting accurate? I thought, based on what it appears that this PR is doing, that the dam was more for Halloween problems during update operations rather than sorting. If that's the case, it also seems like it would be safe to pass the skip and limit along

Reply:
Asking about this comment again. And related to that, whether clearing the skip and limit for dam plans makes sense, or if it should pass down the skip and limit to its child

Reply:
So for a dam, we just need to ensure that the entire inner is consumed before any records are returned. It's not so much about the ordering but the before-after this introduces into the data flow. If the `MemoryDam` is in-memory-limitded, we may actually go over the input multiple times which is weird for a dam. However, in reality we use that plan operator in a way that we are unlimited which in turn ensures that we go over the input exactly once. 

I added that bit of information in the comment.